packages/contracts/contracts/optimistic-ethereum/OVM/bridge/messaging/OVM_L1CrossDomainMessenger.sol

```diff
@@ -155,6 +165,46 @@ contract OVM_L1CrossDomainMessenger is
         emit MessageAllowed(_xDomainCalldataHash);
     }
 
+    function xDomainMessageSender()
+        public
+        override
+        view
+        returns (
+            address
+        )
+    {
+        require(xDomainMsgSender != DEFAULT_XDOMAIN_SENDER, "xDomainMessageSender is not set");
+        return xDomainMsgSender;
+    }
+
+    /**
+     * Sends a cross domain message to the target messenger.
+     * @param _target Target contract address.
+     * @param _message Message to send to the target.
+     * @param _gasLimit Gas limit for the provided message.
+     */
+    function sendMessage(
+        address _target,
+        bytes memory _message,
+        uint32 _gasLimit
+    )
+        override
+        public
+    {
+        // Use the CTC queue length as nonce
+        uint40 queueLength = iOVM_CanonicalTransactionChain(resolve("OVM_CanonicalTransactionChain")).getQueueLength();
+
+        bytes memory xDomainCalldata = Lib_CrossDomainUtils.encodeXDomainCalldata(
+            _target,
+            msg.sender,
+            _message,
+            queueLength
+        );
+
+        _sendXDomainMessage(xDomainCalldata, _gasLimit);
```

Comment:
`resolve("OVM_CanonicalTransactionChain")` is called here, and again in [`_sendXDomainMessage` on line 439](https://github.com/ethereum-optimism/optimism/pull/667/files#diff-78a5e005e505965f63e037a7dc4bf7176c934a405826eb0719ba0b436e2e9134R439). Both require a CALL and an SLOAD. We can probably save gas by caching the address in memory. 
A similar pattern can be seen in the CTC wherever [`queue()`](https://github.com/ethereum-optimism/optimism/blob/master/packages/contracts/contracts/optimistic-ethereum/OVM/chain/OVM_CanonicalTransactionChain.sol#L105-L116) is called.

Reply:
@ben-chain told me that the entire `LibAddressManager` model will be dropped as part of the upgradability work, right Ben? Still I made the suggested changes as they are valid anyhow.